File: 2019/day-1/cpp/src/fuel_counter.cpp
#include "fuel_counter.hpp"

#include <iostream>
#include <fstream>
#include <string>
#include <algorithm>

int fuel_counter(std::vector<int> & masses) {
  int fuel = 0;
  for(std::vector<int>::iterator it = masses.begin();
      it != masses.end(); it++) {
    fuel += fuel_counter(*it);
  }
  return fuel;
}
int fuel_counter(int mass) {
  return std::max((int)(std::floor(mass/3.0)-2),0);
}

int fuel_counter_extended(std::vector<int> & masses) {
  int fuel = 0;
  for(std::vector<int>::iterator it = masses.begin();
      it != masses.end(); it++) {
    fuel += fuel_needed_total(*it);
  }
  return fuel;
}

int fuel_needed_total(int mass) {
  int required_fuel = fuel_counter(mass);
  if (required_fuel > 0)
    return required_fuel + fuel_needed_total(required_fuel);
  return 0;
}

bool readFile(std::string fileName, std::vector<int> & masses) {
  std::ifstream file(fileName);
  if(!file) {
    std::cerr << "Cannot open the File : " << fileName << std::endl;
    return false;
  }

  std::string line;
  while(std::getline(file,line)) {
    if(line.size()>0)
      masses.push_back(std::stoi(line));
  }
  file.close();
  return true;
}

int main() {

  std::vector<int> masses;
  if(readFile("../data.txt",masses)) {
    std::cout<<"Fuel needed for part one: " <<
      fuel_counter(masses) << std::endl;
    std::cout<<"Fuel needed for part one: " <<
      fuel_counter_extended(masses) << std::endl;
  }

  return 0;


  
}

File: 2019/day-1/python/fuel_counter.py
import math

def fuel_counter(masses):
  """ Function that computes the fuel needed for the input masses.
  Expected input is required to either be a list/tuple or scalar"""
  if isinstance(masses,(tuple,list)):
    return sum(fuel_counter(mass) for mass in masses)
  return max(int(math.floor(masses/3)-2),0)

def fuel_needed_total(mass):
  """ Function that computes the fuel needed for the mass and its own fuel.
  Expected input is a scalar """
  required_fuel = fuel_counter(mass)
  if required_fuel > 0:
    return required_fuel + fuel_needed_total(required_fuel)
  return 0

def fuel_counter_extended(masses):
  """ Function that computes the fuel needed for the input masses and the extra fuel needed for the new fuel itself.
  Expected input is required to either be a list/tuple or scalar"""
  if isinstance(masses,(tuple,list)):
    return sum(fuel_counter_extended(mass) for mass in masses)
  return fuel_needed_total(masses)

def main():
  # Open data file and read through all lines
  file_location = "../data.txt"
  try:
    with open(file_location) as f:
      masses = f.readlines()

    # Remove whitespace characters at end of each line
    masses = [int(x.strip()) for x in masses]

    # Print solutions
    print("Fuel needed for part one: " + str(fuel_counter(masses)))
    print("Fuel needed for part two: " + str(fuel_counter_extended(masses)))
  except IOError:
    print("Cannot find file at: " + file_location)

if __name__ == "__main__":
  main()
